crates/noirc_evaluator/src/ssa/node.rs

```diff
@@ -0,0 +1,948 @@
+            }
+            Object::Constants(_) => ObjectType::native_field, //TODO
+            Object::Integer(i) => ObjectType::unsigned(i.num_bits), //TODO signed or unsigned?
+            Object::Linear(_) => {
+                todo!();
+                //ObjectType::native_field
+            }
+            Object::Null => ObjectType::none,
+        }
+    }
+
+    pub fn bits(&self) -> u32 {
+        match self {
+            ObjectType::boolean => 1,
+            ObjectType::native_field => FieldElement::max_num_bits(), //TODO is it the correct value?
+            ObjectType::none => 0,
+            ObjectType::signed(c) => *c,
+            ObjectType::unsigned(c) => *c,
+            //ObjectType::custom(_) => todo!(),
+        }
+    }
+}
+#[derive(Clone, Debug)]
+pub struct Instruction {
+    pub idx: arena::Index,
+    pub operator: Operation,
+    pub rhs: arena::Index,
+    pub lhs: arena::Index,
+    pub res_type: ObjectType, //result type
+    //prev,next: should have been a double linked list so that we can easily remove an instruction during optimisation phases
+    pub parent_block: arena::Index,
+    pub is_deleted: bool,
+    pub res_name: String,
+    pub bit_size: u32, //TODO only for the truncate instruction...: bits size of the max value of the lhs.. a merger avec ci dessous!!!TODO
+    pub max_value: BigUint, //TODO only for sub instruction: max value of the rhs
+
+    //temp: todo phi subtype
+    pub phi_arguments: Vec<(arena::Index, arena::Index)>,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum NodeEval {
+    Const(FieldElement, ObjectType),
+    Idx(arena::Index),
+}
+
+impl NodeEval {
+    pub fn to_const_value(self) -> Option<FieldElement> {
+        match self {
+            NodeEval::Const(c, _) => Some(c),
+            _ => None,
+        }
+    }
+    pub fn to_index(self) -> Option<arena::Index> {
+        match self {
+            NodeEval::Idx(i) => Some(i),
+            NodeEval::Const(_, _) => None,
+        }
+    }
+}
+
+impl Instruction {
+    pub fn new(
+        op_code: Operation,
+        lhs: arena::Index,
+        rhs: arena::Index,
+        r_type: ObjectType,
+        parent_block: Option<arena::Index>,
+    ) -> Instruction {
+        let id0 = crate::ssa::code_gen::IRGenerator::dummy_id();
+        let p_block = parent_block.unwrap_or(id0);
+        Instruction {
+            idx: id0,
+            operator: op_code,
+            lhs,
+            rhs,
+            res_type: r_type,
+            res_name: String::new(),
+            is_deleted: false,
+            parent_block: p_block,
+            bit_size: 0,
+            max_value: BigUint::zero(),
+            phi_arguments: Vec::new(),
+        }
+    }
+
+    pub fn print_i(&self) -> String {
+        if self.res_name.is_empty() {
+            format!("{:?}", self.idx)
+        } else {
+            self.res_name.clone()
+        }
+    }
+
+    //indicates if the operation is a substraction
+    pub fn is_sub(&self) -> bool {
```

Comment:
Can you add a comment stating why this particular opcode needs this check? Seems like it belongs in the integer file?